### DISCUSSION TITLE
Install Custom Theme


Hi Guys! Does anyone here have their custom themes.Would you mind sharing it here?

### DISCUSSION THREAD

**Discussion by kRHYME7** [prasanthrangan/hyprdots#468] [Show and tell]
I wrote a quick script to install/patch a custom Theme. 

First is compile everything into a single folder, we can name it ``` Fav-Theme ```
The instructions on making a theme are just the same with https://github.com/prasanthrangan/hyprdots/wiki/Theming
Here is the file tree 
/path/to/custom/Fav-Theme

├── Arcs
│   ├── Cursor_Fav-Theme.tar.gz
│   ├── Gtk_Fav-Theme.tar.gz
│   └── Icon_Fav-Theme.tar.gz
├── hypr
│   └── themes
│       └── Fav-Theme.conf
├── import.lst
├── kitty
│   └── themes
│       └── Fav-Theme.conf
├── Kvantum
│   └── Fav-Theme
│       ├── Fav-Theme.kvconfig
│       └── Fav-Theme.svg
├── qt5ct
│   └── colors
│       └── Fav-Theme.conf
├── rofi
│   └── themes
│       └── Fav-Theme.rasi
├── swww
│   ├── Fav-Theme
│   │   ├── wall-1.jpg
│   │   ├── wall-2.jpeg
│   │   ├── wall..jpg
│   │   ├── -wallpaper-11-scaled.jpg
│   │   ├── Fav-Theme-wallpaper-9-scaled.jpg
│   │   └── anywallpaper.jpg
│   └── wall.ctl
└── waybar
    └── themes
        └── Fav-Theme.css

then this ``` import.lst ```  are the Icons, cursors, and GTK things that you use and the path where they should be added.
```
❯ cat ./import.lst
Gtk_Fav-Theme|/usr/share/themes
Gtk_Fav-Theme|$HOME/.themes
Icon_Fav-Theme|/usr/share/icons
Icon_Fav-Theme|$HOME/.icons
Cursor_Fav-Theme|/usr/share/icons
Cursor_Fav-Theme|$HOME/.icons%      
```

Also in this way, the Compressed "Themepack" can be hosted on your repo. 
This script is on testing so can I try your custom themes?

Here is the script, This one is a very messy script but functional. 
https://github.com/kRHYME7/KhingDom/blob/Khingdom/Configs/.config/hypr/scripts/themeinstall.sh

**Comment by T-Crypt** [comment]
I'll test this, and work on the function as well. 

I see you added my function for the xz file compression already! Very nice ❤️

Let me add a function to ask the user if they are using a custom theme directory with everything or if they would like to enter the GitHub URL and branch to their theme then it'll clone and run the respective copy and archive jobs.

**Reply by kRHYME7** [reply]
Yes Please

**Reply by prasanthrangan** [reply]
Hey @kRHYME7 , your script looks good...
you can simplify it by reusing the existing `restore_cfg.sh`

for example your `themepatcher.sh` script could receive the theme name (`Fav-Theme`) and git repo link as parameter 
```
./themepatcher.sh Fav-Theme https://github.com/user/Fav-Theme-branch
```


1. then clone the repo and override it to `$CfgDir` path in `restore_cfg.sh`
2. validate the files cloned, it should have these files -
  ```
  ~/.config/hypr/themes/Fav-Theme.conf
  ~/.config/kitty/themes/Fav-Theme.conf
  ~/.config/Kvantum/Fav-Theme/Fav-Theme.kvconfig
  ~/.config/Kvantum/Fav-Theme/Fav-Theme.svg
  ~/.config/qt5ct/colors/Fav-Theme.conf
  ~/.config/rofi/themes/Fav-Theme.rasi
  ~/.config/swww/Fav-Theme/wallapers*
  ~/.config/waybar/themes/Fav-Theme.css
  ```
3. generate `Fav-Theme_restore_cfg.lst` required for `restore_cfg.sh` to run as 

```
Y|${HOME}/.config/hypr/themes|Fav-Theme.conf|hyprland
Y|${HOME}/.config/kitty/themes|Fav-Theme.conf|kitty
Y|${HOME}/.config/Kvantum|Fav-Theme|kvantum
Y|${HOME}/.config/qt5ct/colors|Fav-Theme.conf|qt5ct
Y|${HOME}/.config/rofi/themes|Fav-Theme.rasi|rofi
N|${HOME}/.config/swww|Fav-Theme|swww
Y|${HOME}/.config/waybar/themes|Fav-Theme.css|waybar
```
4. override the `restore_cfg.lst` loop with the generated `Fav-Theme_restore_cfg.lst` file
5. execute `restore_cfg.sh` with the overrides
6. check and add theme entry in `~/.config/swww/wall.ctl`

**Reply by kRHYME7** [reply]
Okies, I should integrate this

**Reply by kRHYME7** [reply]
Also I will try to grasp how the new restore script works haha

**Reply by kRHYME7** [reply]
@Tittu I think I get it now, But by doing so I will modify the ./restore_cfg.sh right? 
Is it cool if I will somehow write it that if 
$1 and $2 exist as parameters(Theme-name and valid Github repo ) I will execute those things mentioned above,  if $1 and $2 were null then just the normal ./restore_cfg.sh.  


If I understood this right:
 The git Repo link is a fork of hyprdots, right?  "Fav-Theme" is the keyword to validate and search for the required files for the theme. If validation is successful we can now generate the ``` Fav-Theme_restore_cfg.lst ``` and use ./restore_cfg.sh to restore all theme related configs mentioned inside ```` Fav-Theme_restore_cfg.lst ``` 
~~But doing so will override the default hyprdots configs installed right? Or it will be added?~~

Edit: I get it now It should Append it because Fav-Theme ``` was explicitly added  Thanks !!!
 ~~My primary Goal for the script is to make it install and append a theme from another branch to your branch so that I could still have the bleeding edge changes from the main branch. One of the cool things in hyprdots is the seamless changing of themes etc.~~

I can clone a branch and compile only the mentioned  ``` Fav-Theme ``` configs and then install it alongside your hyprdots default themes. Then if this is implemented well we can add a function at the end of the install script to install their preferred themes
 ``` ./themepatcher.sh Fav-Theme https://github.com/user/Fav-Theme-branch ```

What is your recommended path for the clone repo? Is the ~/.cache/hyprdots already or soon to be implemented? 


~~Can I ask what workflow you are considering implementing regarding patching a  theme? Should it be set up before the ``` ./install.sh -r ``` hyprdots repo OR install the hyprdots repo and then patch the Fav-Theme?~~

**Reply by prasanthrangan** [reply]
> @tittu I think I get it now, But by doing so I will modify the ./restore_cfg.sh right?
> Is it cool if I will somehow write it that if
> $1 and $2 exist as parameters(Theme-name and valid Github repo ) I will execute those things mentioned above, if $1 and $2 were null then just the normal ./restore_cfg.sh.

I've modified restore_cfg.sh for you, here you go - 
[restore_cfg.sh.txt](https://github.com/prasanthrangan/hyprdots/files/13321028/restore_cfg.sh.txt)

so the `./themepatcher.sh` can clone the theme  repo (not required to be a fork, it just needs the theme files) to `$HOME/Clone/Fav-Theme` and then it should execute `restore_cfg.sh` as 
```
./restore_cfg.sh "Fav-Theme" "$HOME/Clone/Fav-Theme"
```

**Reply by kRHYME7** [reply]
Thanks! Do you find a way to download a specific folder like a wallpaper? 
I managed to extract all config files except wallpaper.

Cloning a  repo takes time and storage.

**Reply by prasanthrangan** [reply]
you can either clone it or use local directory... for example -
```
./themepatcher.sh Fav-Theme https://github.com/user/Fav-Theme-branch #will clone it to $HOME/Clone/Fav-Theme

./themepatcher.sh Fav-Theme /path/to/theme/dir #directory should have the config files for Fav-Theme
```


regarding storage use  `git clone --depth 1 https://github.com/user/theme-repo`
you can keep the wallpapers in the same repo.

**Reply by kRHYME7** [reply]
https://github.com/kRHYME7/KhingDom/blob/Khingdom/Scripts/themepatcher.sh 
I'll refactor this tomorrow.

**Reply by prasanthrangan** [reply]
> https://github.com/kRHYME7/KhingDom/blob/Khingdom/Scripts/themepatcher.sh 
> I'll refactor this tomorrow.

To simplify it , the theme repo just needs these files 

  "Configs/.config/hypr/themes/$Fav_Theme.conf"
  "Configs/.config/kitty/themes/$Fav_Theme.conf"
  "Configs/.config/Kvantum/$Fav_Theme/$Fav_Theme.kvconfig"
  "Configs/.config/Kvantum/$Fav_Theme/$Fav_Theme.svg"
  "Configs/.config/qt5ct/colors/$Fav_Theme.conf"
  "Configs/.config/rofi/themes/$Fav_Theme.rasi"
  "Configs/.config/swww/$Fav_Theme/"
  "Configs/.config/waybar/themes/$Fav_Theme.css"

And then tar files for gtk theme, icon (optional), vscode extension..

**Reply by kRHYME7** [reply]
I will issue a PR so we can have a better Workspace
